File: 백준/Gold/16724. 피리 부는 사나이/피리 부는 사나이.cc
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    int n, m; cin >> n >> m;
    vector<string> board(n);
    vector<vector<int>> vis(n , vector<int>(m, 0));
    for(int i = 0; i < n; i++) cin >> board[i];
    int ans, idx; ans = idx = 0;
    for(int i = 0; i < n; i++) for(int j = 0; j < m; j++){
        if(vis[i][j]) continue;
        idx++;
        int x = i, y = j;
        while(!vis[x][y]){
            vis[x][y] = idx;
            if(board[x][y] == 'U') x--;
            else if(board[x][y] == 'D') x++;
            else if(board[x][y] == 'L') y--;
            else y++;
        }
        if(vis[x][y] == idx) ans++;
    }
    cout << ans;
    return 0;
}

File: 백준/Gold/16724. 피리 부는 사나이/피리 부는 사나이.py
import sys
input = lambda: sys.stdin.readline().rstrip()
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)

def main():
    n, m = map(int, input().split())
    board = [input() for _ in range(n)]
    vis = [[0]*m for _ in range(n)]
    ans = idx = 0
    move = {'U': (-1, 0), 'D': (1, 0), 'L': (0, -1), 'R': (0, 1)}
    for i in range(n):
        for j in range(m):
            if vis[i][j]: continue
            idx += 1
            x, y = i, j
            while not vis[x][y]:
                vis[x][y] = idx
                dx, dy = move[board[x][y]]
                x, y = x+dx, y+dy
            if vis[x][y] == idx: ans += 1
    print(ans)
    return
if __name__ == "__main__":
    main()